dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorRecordStore.cs

```diff
@@ -0,0 +1,234 @@
+    /// Initializes a new instance of the <see cref="PineconeVectorRecordStore{TRecord}"/> class.
+    /// </summary>
+    /// <param name="pineconeClient">Pinecone client that can be used to manage the collections and vectors in a Pinecone store.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    /// <exception cref="ArgumentNullException">Thrown if the <paramref name="pineconeClient"/> is null.</exception>
+    /// <exception cref="ArgumentException">Thrown for any misconfigured options.</exception>
+    public PineconeVectorRecordStore(Sdk.PineconeClient pineconeClient, PineconeVectorRecordStoreOptions<TRecord>? options = null)
+    {
+        Verify.NotNull(pineconeClient);
+
+        this._pineconeClient = pineconeClient;
+        this._options = options ?? new PineconeVectorRecordStoreOptions<TRecord>();
+
+        if (this._options.MapperType == PineconeRecordMapperType.PineconeVectorCustomMapper)
+        {
+            if (this._options.VectorCustomMapper is null)
+            {
+                throw new ArgumentException($"The {nameof(PineconeVectorRecordStoreOptions<TRecord>.VectorCustomMapper)} option needs to be set if a {nameof(PineconeVectorRecordStoreOptions<TRecord>.MapperType)} of {nameof(PineconeRecordMapperType.PineconeVectorCustomMapper)} has been chosen.", nameof(options));
+            }
+
+            this._mapper = this._options.VectorCustomMapper;
+        }
+        else
+        {
+            this._mapper = new PineconeVectorStoreRecordMapper<TRecord>(new PineconeVectorStoreRecordMapperOptions
+            {
+                VectorStoreRecordDefinition = this._options.VectorStoreRecordDefinition
+            });
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<TRecord?> GetAsync(string key, GetRecordOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(key);
+
+        var records = await this.GetBatchAsync([key], options, cancellationToken).ToListAsync(cancellationToken).ConfigureAwait(false);
+
+        return records.FirstOrDefault();
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<TRecord> GetBatchAsync(
+        IEnumerable<string> keys,
+        GetRecordOptions? options = default,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(keys);
+
+        var collectionName = this.ChooseCollectionName(options?.CollectionName);
+        var indexNamespace = this.GetIndexNamespace();
+        var mapperOptions = new StorageToDataModelMapperOptions { IncludeVectors = options?.IncludeVectors ?? false };
+
+        var index = await this.GetIndexAsync(collectionName, cancellationToken).ConfigureAwait(false);
+
+        var results = await RunOperationAsync(
+            collectionName,
+            UpsertOperationName,
+            () => index.Fetch(keys, indexNamespace, cancellationToken)).ConfigureAwait(false);
```

Comment:
The goal for the abstraction at the moment is to return the subset that was found. So, only found records will be returned, and the resultset may therefore be smaller than the requested keys.
Is that also the behavior of the fetch operation?

Reply:
Yes, basically returning vectors by Id (and returning all that were found), as opposed Query operation which does similarity search. I added a test where one of the ids doesn't exist to verify it's still a success.